toolchain/check/generic.cpp

```diff
@@ -490,38 +503,147 @@ auto BuildGenericDecl(Context& context, SemIR::InstId decl_id)
+// element in the given eval block. Otherwise returns the ID unchanged.
+static auto ReattachConstant(Context& context, SemIR::GenericId generic_id,
+                             llvm::ArrayRef<SemIR::InstId> eval_block,
+                             SemIR::ConstantId constant_id)
+    -> SemIR::ConstantId {
+  if (!constant_id.has_value() || !constant_id.is_symbolic()) {
+    return constant_id;
+  }
+
+  auto& symbolic_const =
+      context.constant_values().GetSymbolicConstant(constant_id);
+  if (symbolic_const.generic_id != generic_id) {
+    // Constant doesn't refer into this generic.
+    return constant_id;
+  }
+
+  CARBON_CHECK(
+      symbolic_const.index.region() == SemIR::GenericInstIndex::Declaration,
+      "Definition region of redeclaration should not be referenced");
+  return context.constant_values().GetAttached(
+      eval_block[symbolic_const.index.index()]);
+}
+
+// Same as `ReattachConstant` but for a type.
+static auto ReattachType(Context& context, SemIR::GenericId generic_id,
+                         llvm::ArrayRef<SemIR::InstId> eval_block,
+                         SemIR::TypeId type_id) -> SemIR::TypeId {
+  return context.types().GetTypeIdForTypeConstantId(ReattachConstant(
+      context, generic_id, eval_block, context.types().GetConstantId(type_id)));
+}
+
 auto FinishGenericRedecl(Context& context, SemIR::GenericId generic_id)
     -> void {
   if (!generic_id.has_value()) {
-    context.generic_region_stack().Pop();
+    DiscardGenericDecl(context);
     return;
   }
 
-  auto definition_block_id = MakeGenericEvalBlock(
-      context, generic_id, SemIR::GenericInstIndex::Region::Declaration,
-      /*inside_redeclaration=*/true);
-  CARBON_CHECK(definition_block_id == SemIR::InstBlockId::Empty);
+  // Find the old and new eval blocks.
+  auto old_eval_block_id =
+      context.generics()
+          .Get(generic_id)
+          .GetEvalBlock(SemIR::GenericInstIndex::Declaration);
+  CARBON_CHECK(old_eval_block_id.has_value(),
+               "Old generic is not fully declared");
+
+  auto old_eval_block = context.inst_blocks().Get(old_eval_block_id);
+  auto new_eval_block = context.generic_region_stack().PeekEvalBlock();
+
+  // Check the eval blocks are computing the same constants in the same order.
+  // This should always be the case because we have already verified they have
+  // the same parse tree, and the poisoning rules mean that all entities they
+  // refer to are also the same.
+  //
+  // Note that it's OK if the first difference is that an old instruction has no
+  // corresponding new instruction; we wouldn't have used that anyway. This
+  // happens for `ImplDecl`, for which the witness is included in the eval block
+  // of the first declaration.
+  if (auto [old_inst_id, new_inst_id] = FirstDifferenceBetweenEvalBlocks(
+          context, old_eval_block, new_eval_block);
+      new_inst_id.has_value()) {
+    context.TODO(new_inst_id,
```

Comment:
Maybe CARBON_FATAL?

Reply:
I agree printing things can be helpful in the case of an error. Maybe a context.UNEXPECTED() would fit for cases like this a bit better than context.TODO(), but no need to do that now. Putting a FATAL after to TODOs isn't as helpful as might sound because then when running a file test it won't rewrite the file so you will have a hard time ever seeing the printed diagnostics. So it's okay to leave as is I think, but maybe we can think about changing the naming of TODO since that looks misleading. If you agree, maybe drop a TODO on the TODOs (^_^') about renaming them?